libpod/image/pull.go

```diff
@@ -198,12 +200,74 @@ func (ir *Runtime) pullGoalFromImageReference(ctx context.Context, srcRef types.
 
 // pullImageFromHeuristicSource pulls an image based on inputName, which is heuristically parsed and may involve configured registries.
 // Use pullImageFromReference if the source is known precisely.
-func (ir *Runtime) pullImageFromHeuristicSource(ctx context.Context, inputName string, writer io.Writer, authfile, signaturePolicyPath string, signingOptions SigningOptions, dockerOptions *DockerRegistryOptions, label *string) ([]string, error) {
+func (ir *Runtime) pullImageFromHeuristicSource(ctx context.Context, config *NewImageConfig) ([]string, error) {
 	span, _ := opentracing.StartSpanFromContext(ctx, "pullImageFromHeuristicSource")
 	defer span.Finish()
+	sc := GetSystemContext(config.SignaturePolicyPath, config.Authfile, false)
+	goal, err := ir.getPullGoalFromHeuristicSource(ctx, sc, config.Name)
+	if err != nil {
+		return nil, err
+	}
+	return ir.doPullImage(ctx, sc, *goal, config.Writer, config.SigningOptions, config.DockerOptions, config.Label)
+}
+
+// pullImageFromHeuristicSourceIfNecessary checks an image is out of date with its source,
+// and if so, pulls an updated image.
+// Rather than search for the name in config, use the names of localImg, so we don't have to
+// refind the transport and reconstruct the correct name.
+func (ir *Runtime) pullImageFromHeuristicSourceIfNecessary(ctx context.Context, localImg *storage.Image, config *NewImageConfig) ([]string, error) {
+	span, _ := opentracing.StartSpanFromContext(ctx, "pullImageFromHeuristicSourceIfNecessary")
+	defer span.Finish()
+
+	var inspectErrors *multierror.Error
 
+	sc := GetSystemContext(config.SignaturePolicyPath, config.Authfile, false)
+	var goal *pullGoal
+	for _, name := range localImg.Names {
+		goal, _ = ir.getPullGoalFromHeuristicSource(ctx, sc, name)
```

Comment:
`Image.Names` contains values suitable for `reference.ParseNormalizedNamed`; it can’t contain a _transport_`:`_image_-formatted name, so the generality of `getPullGoalFromHeuristicSource` is unnecessary (and, in principle, incorrect).

_If_ this is going to use `localImg.Names` as the source of information, this should ideally use the `decomposedImage.hasRegistry` case of `pullGoalFromPossiblyUnqualifiedName` (probably extract that into a `pullGoalFromQualifiedDockerReference` helper, or in the worst case just call `pullGoalFromPossiblyUnqualifiedName` although we do know that the name is qualified).

All of this is not going to work for images which come from non-`docker://` sources at all (most cases handled by `pullGoalFromImageReference`, though, and AFAIK we don’t record the actual origin of images in a way that would make that possible.

(Yet again, the `!forcePull` code path has always worked only with docker-like names, so maybe this does not matter _that_ much?)

See also the note about using the creation timestamp, and the note about “found in any registry”.